### PR TITLE
feat(template-operator): init declaritive pattern

### PR DESCRIPTION
**Pull request by adityabhatia**
depends on: https://github.com/kyma-project/manifest-operator/pull/67
Update `go.mod` once the above PR is merged

### CODE REVIEW

File: samples/template-operator/operator/controllers/sample_controller.go [hunk 1]
```diff
@@ -47,188 +39,16 @@ const deletionFinalizer = "deletion-finalizer"
...
-	// TODO: ready logic here
-
-	// Example: If Ready state, check consistency of deployed module
-	return nil
-}
-
 // SetupWithManager sets up the controller with the Manager.
 func (r *SampleReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	r.Config = mgr.GetConfig()
+	if err := r.Inject(mgr, &v1alpha1.Sample{},
```
Comment (jakobmoellersap) [inline samples/template-operator/operator/controllers/sample_controller.go:45]: Why use Inject as a pattern here? You could simply abstract away the reconciliation with controller-runtime no?

E.G. in the declarative Example it shows 

```
func (r *GuestbookReconciler) setupReconciler(mgr ctrl.Manager) error {
	labels := map[string]string{
		"example-app": "guestbook",
	}

	r.watchLabels = declarative.SourceLabel(mgr.GetScheme())

	return r.Reconciler.Init(mgr, &api.Guestbook{},
		declarative.WithObjectTransform(declarative.AddLabels(labels)),
		declarative.WithOwner(declarative.SourceAsOwner),
		declarative.WithLabels(r.watchLabels),
		declarative.WithStatus(status.NewBasic(mgr.GetClient())),
		declarative.WithApplyPrune(),
		declarative.WithReconcileMetrics(0, nil),
	)
}

func (r *GuestbookReconciler) SetupWithManager(mgr ctrl.Manager) error {
	if err := r.setupReconciler(mgr); err != nil {
		return err
	}

	c, err := controller.New("guestbook-controller", mgr, controller.Options{Reconciler: r})
	if err != nil {
		return err
	}

	// Watch for changes to Guestbook
	err = c.Watch(&source.Kind{Type: &api.Guestbook{}}, &handler.EnqueueRequestForObject{})
	if err != nil {
		return err
	}

	// Watch for changes to deployed objects
	_, err = declarative.WatchChildren(declarative.WatchChildrenOptions{Manager: mgr, Controller: c, Reconciler: r, LabelMaker: r.watchLabels})
	if err != nil {
		return err
	}

	return nil
}
```

I think we should make the setupWithManager a bit cleaner, still it looks great compared to before!!!

File: samples/template-operator/operator/controllers/sample_controller.go [hunk 2]
```diff
@@ -47,188 +39,16 @@ const deletionFinalizer = "deletion-finalizer"
...
+		declarative.WithResourceLabels(
+			map[string]string{"sampleKey": "sampleValue"},
+		)); err != nil {
+		return err
```
Comment (jakobmoellersap) [inline samples/template-operator/operator/controllers/sample_controller.go:49]: err check not needed

File: samples/template-operator/operator/controllers/sample_controller.go (hunk 1 above)
Comment (adityabhatia) [reply samples/template-operator/operator/controllers/sample_controller.go:52]: Did not quite understand the abstracting with controller-runtime part.

File: samples/template-operator/operator/controllers/sample_controller.go (hunk 2 above)
Comment (adityabhatia) [reply samples/template-operator/operator/controllers/sample_controller.go:58]: Why?

Comment (jakobmoellersap) [reply samples/template-operator/operator/controllers/sample_controller.go:53]: Actually not sure anymore, sorry for the confusion

**Review by ruanxin — APPROVED** [review]

**Pull request merged**